internal/datastore/spanner/gc.go

```diff
@@ -0,0 +1,66 @@
+package spanner
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	sq "github.com/Masterminds/squirrel"
+	"github.com/go-co-op/gocron"
+	"github.com/rs/zerolog/log"
+)
+
+type cancelFunc func()
+
+var noopCancelFunc cancelFunc = func() {}
+
+func (sd spannerDatastore) runGC() cancelFunc {
+	if sd.config.gcInterval <= 0 {
+		log.Info().Stringer("interval", sd.config.gcInterval).Msg("garbage collection: disabled")
+		return noopCancelFunc
+	}
+
+	log.Info().Stringer("interval", sd.config.gcInterval).Msg("garbage collection: starting")
+
+	s := gocron.NewScheduler(time.UTC)
+	var numRemoved int64
+	_, err := s.Every(sd.config.gcInterval).Do(func() {
+		ctx, span := tracer.Start(context.Background(), "CollectGarbage")
+		defer span.End()
+
+		ctx, cancel := context.WithTimeout(ctx, sd.config.gcInterval)
+		defer cancel()
+
+		spannerNow, err := sd.now(ctx)
+		if err != nil {
+			log.Error().Err(err).Msg("garbage collection: error computing datastore time")
+		}
+
+		oldestRevision := spannerNow.Add(-1 * sd.config.gcWindow)
+
+		stmt, args, err := sql.Delete(tableChangelog).Where(sq.Lt{colChangeTS: oldestRevision}).ToSql()
+		if err != nil {
+			log.Error().Err(err).Msg("garbage collection: error creating delete statement")
+		}
+
+		_, err = sd.client.ReadWriteTransaction(ctx, func(ctx context.Context, rwt *spanner.ReadWriteTransaction) error {
+			numRemoved, err = rwt.Update(ctx, statementFromSQL(stmt, args))
+			return err
+		})
+		if err != nil {
+			log.Error().Err(err).Msg("garbage collection: error deleting entries")
+		}
+
+		log.Info().Int64("removed", numRemoved).Stringer("before", oldestRevision).
+			Msg("garbage collection: removed changelog entries")
+	})
+	if err != nil {
+		panic(fmt.Errorf("unable to start garbage collection: %w", err))
```

Comment:
instead of panicing, this should thread this back up as an error so that the other goroutines can be cancelled / have a chance to gracefully stop

Reply:
+1

Reply:
Output after doing this:
```
9:46AM INF set log level new level=debug
9:46AM INF set tracing provider new provider=none
9:46AM INF using spanner datastore engine
9:46AM INF gc configured with jitter factor=0.2
9:46AM INF garbage collection: starting interval=3m24.678454111s
9:46AM INF request hedging enabled hedgingQuantile=0.95 initialSlowRequest=10ms maxRequests=1000000
9:46AM DBG configured combined dispatcher upstream=
9:46AM WRN grpc server serving plaintext prefix=grpc
9:46AM INF grpc server started listening addr=:50051 network=tcp prefix=grpc
9:46AM WRN http server serving plaintext addr=:9090 prefix=metrics
9:46AM WRN http server serving plaintext addr=:8080 prefix=dashboard
^C9:46AM INF received interrupt
9:46AM INF shutting down
9:46AM INF http server stopped serving addr=:8080 prefix=dashboard
9:46AM INF http server stopped serving addr=:9090 prefix=metrics
9:46AM INF grpc server stopped listening addr=:50051 network=tcp prefix=grpc
9:46AM INF garbage collection: stopping
```